### ISSUE TITLE
Use QNames as anchors in generated HTML

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#2604]
Rather than line numbers, use qualified names as anchors in `<a name="...">`.

**Comment by andreasabel** [comment]
Testing the generated html could maybe done with https://github.com/gjtorikian/html-proofer

**Comment by andreasabel** [comment]
CHANGELOG documentation of this enhancement:

  The new, symbolic anchors look like
  ```html
  <a name="test1">
  <a name="M.bla">
  ```
  while the old anchors just give the character position in the file:
  ```html
  <a name="42">
  ```

  Top-level module names do not get a symbolic anchor, since the position of
  a top-level module is defined to be the beginning of the file.

  Example:

  ```agda
  module Issue2604 where   -- Character position anchor

  test1 : Set₁             -- Issue2604.html#test1
  test1 = bla
    where
    bla = Set              -- Character position anchor

  test2 : Set₁             -- Issue2604.html#test2
  test2 = bla
    where
    bla = Set              -- Character position anchor

  test3 : Set₁             -- Issue2604.html#test3
  test3 = bla
    module M where         -- Issue2604.html#M
    bla = Set              -- Issue2604.html#M.bla

  module NamedModule where -- Issue2604.html#NamedModule
    test4 : Set₁           -- Issue2604.html#NamedModule.test4
    test4 = M.bla

  module _ where           -- Character position anchor
    test5 : Set₁           -- Character position anchor
    test5 = M.bla
  ```

**Comment by gallais** [comment]
:champagne:  This is something I wish we'd had for a long time! Urls will now be more stable even after refactorings (they will be kept the same as long as the toplevel definitions have kept the same name / stayed in the same module).

A few remarks:

* What about unicode identifiers? It might be hard to write these urls by hand
* What about mixfix identifiers? I suppose we get underscore in the url? Which is fine.
* What about identifiers containing hashes? I suppose that ought to break the anchor mechanism!

**Comment by andreasabel** [comment]
Unicode identifiers will be unicode anchors.  No problem from the browser side, as far as I can see.  
From the user side: GIGO (garbage in, garbage out).  You reap what you sow.
But I do not imagine users to type these anchors into the browser.  It is for linking from a PDF, and there you can do probably enough on the LaTeX side to ease the pain typing unicode anchors.  We will see, I have not used it in production yet.

Hashes might indeed be bad.  Either we apply GIGO here as well, or we fall back to character position anchors should there be hashes in the name.

**Comment by nad** [comment]
I think we should try to make sure that the generated HTML is correct. The files generated by an older version of Agda seem to be valid according to the W3C validators ([HTML](https://validator.w3.org/), [CSS](https://jigsaw.w3.org/css-validator/)). However, perhaps the old anchor names, containing nothing but digits, are incorrect. For some discussion, see https://github.com/haskell/haddock/issues/45.

**Comment by fredrikNordvallForsberg** [comment]
If I understand the XHTML spec correctly, there are basically no restrictions on the id attribute values, except that [it must be nonempty and not contain any spaces](http://w3c.github.io/html/dom.html#element-attrdef-global-id) (which will never be a problem), and [the characters " ' & < should be entity-encoded](https://stackoverflow.com/questions/866706/which-characters-are-invalid-unless-encoded-in-an-xml-attribute). (Note that HTML 4 requires that [the identifier starts with a letter](https://mathiasbynens.be/notes/html5-id-class) but we are generating XHTML, not HTML 4.)

If we [URI-encode](https://en.wikipedia.org/wiki/Percent-encoding) the href URIs (in particular "#" becomes "%23"), in theory everything should then be fine? (Arguably it should be up to the [xhtml](https://github.com/haskell/xhtml) package to do the right encoding... It already includes [a function for entity-encoding the illegal characters above](http://hackage.haskell.org/package/xhtml-3000.2.1/docs/Text-XHtml-Strict.html#v:stringToHtmlString).)

**Comment by nad** [comment]
> in theory everything should then be fine?

In practice the generated HTML might be read by, say, some anonymous reviewer, using a browser that is not standards-compliant, so I think we should avoid corner cases, and test using several common browsers.

**Comment by andreasabel** [comment]
I tested `#` in anchors in Firefox 54.0, and it worked fine without URI encoding.
Also on Chromium 58.0.
If you have other browsers, please report your test results here.
Try it out on https://rawgit.com/agda/agda/master/test/LaTeXAndHTML/succeed/Issue2604.html

**Comment by andreasabel** [comment]
For now, I do not intend to change anything.  Closing.  Feel free to reopen if there are problems.

**Comment by nad** [comment]
The W3C validator [complains](https://validator.w3.org/check?uri=https%3A%2F%2Frawgit.com%2Fagda%2Fagda%2Fmaster%2Ftest%2FLaTeXAndHTML%2Fsucceed%2FIssue2604.html&charset=%28detect+automatically%29&doctype=Inline&group=0).

**Comment by gallais** [comment]
> Arguably it should be up to the xhtml package to do the right encoding...

I've tried porting the html backend to the blaze-html package and the same issue shows up: links are not url encoded either...

A popular solution seems to be to use [HTTP](https://hackage.haskell.org/package/HTTP)'s `urlencode` but that's a big dependency to add for something as simple as that. I've come across [uri-encode](https://hackage.haskell.org/package/uri-encode) and have tried it. It does percent encoding perfectly fine but...

https://validator.w3.org is not happy about the encoded `name`s:

> ⨂ <a name="%23" href="Issue2604.html#%23" class="Function"
Line 173, Column 10: character "%" is not allowed in the value of attribute "name"
It is possible that you violated the naming convention for this attribute. For example, id and name attributes must begin with a letter, not a digit.

So, to recap: in XHTML 1.0 Strict, `#` is illegal in names and the uri-encoded version is also rejected because `%` is illegal too!

**Comment by nad** [comment]
Shouldn't this issue be fixed (or the changes reverted) before 2.5.3 is released? I don't think we should break links now, and then again when the next version of Agda is released.

**Comment by andreasabel** [comment]
Is there evidence for breakage?

A yes, and a paper submission depends on this feature, so I would be glad if it was released soon.

**Comment by nad** [comment]
If you intend to have links to identifiers in published papers, then it seems reasonable to me to make the links as stable as possible.

I gave you evidence of breakage above: we generate syntactically incorrect code.

I expect that it is easy to fix this by using another encoding scheme than the one @gallais tried above, for instance the [z-encoding used by GHC](https://ghc.haskell.org/trac/ghc/wiki/Commentary/Compiler/SymbolNames) (but perhaps something that preserves more characters).

**Comment by andreasabel** [comment]
I guess the current popular browsers skip the syntax check since nothing is gained by the restriction it imposes.

One thing to keep in mind is that the anchor names should be as similar to the QNames as possible, since the user writes them in his LaTeX document.  If we do some encoding, we should also ship a LaTeX macro that performs this encoding, to keep things simple to use.

The current use is, for instance:
```latex
\newcommand{\developURL}[2]{html/#1\#\detokenize{#2}}
...
\begin{linklemma}{\developURL{Definition.Typed.Weakening.html}{_•ₜ_}}[Weakening composition]
...
```
(The LaTeX function `\detokenize` does not do much coding.)

**Comment by nad** [comment]
> If we do some encoding, we should also ship a LaTeX macro that performs this encoding, to keep things simple to use.

We don't have to do this before the next release, the LaTeX macro can be added later. However, we can keep it in mind when deciding what kind of encoding scheme to use.

**Comment by nad** [comment]
> If I understand the XHTML spec correctly, there are basically no restrictions on the id attribute values

We are generating `name` attributes.

**Comment by nad** [comment]
I've read a small part of the specification of [XHTML 1.0](http://www.w3.org/TR/2002/REC-xhtml1-20020801). Some notes:
* The `id` attribute [must be used](https://www.w3.org/TR/xhtml1/#h-4.10) to specify anchor targets.
* For compatibility reasons it is [recommended](https://www.w3.org/TR/xhtml1/#C_8) to use both `name` and `id` attributes, with equal attribute values.
* The `name` attributes [can take all values](https://www.w3.org/TR/xhtml1/#C_8) that `id` attributes can take.
* For compatibility reasons these attributes should [match the pattern `[A-Za-z][A-Za-z0-9:_.-]*`](https://www.w3.org/TR/xhtml1/#C_8).

**Comment by nad** [comment]
On the other hand, perhaps we can ignore compatibility guidelines from 2002.

I replaced all `name` attributes with `id` in the test case mentioned above. The validator complains about attribute values starting with digits, as well as attributes containing `#`, `%` and `∀`.

> If I understand the XHTML spec correctly, there are basically no restrictions on the id attribute values, except that [it must be nonempty and not contain any spaces](http://w3c.github.io/html/dom.html#element-attrdef-global-id) (which will never be a problem)

Here you have referred to a draft of the HTML 5.3 standard. I suggest that we switch to HTML 5. While the current HTML 5 standard [specifies](https://www.w3.org/TR/2014/REC-html5-20141028/dom.html#the-id-attribute) that the only restrictions are that the values must be non-empty and that [space characters](https://www.w3.org/TR/2014/REC-html5-20141028/infrastructure.html#space-character) may not be used, `href` attributes [must contain](https://www.w3.org/TR/2014/REC-html5-20141028/links.html#links-created-by-a-and-area-elements) [valid URLs](https://www.w3.org/TR/2014/REC-html5-20141028/infrastructure.html#valid-url) (optionally surrounded by some kind of whitespace). I guess that the names can contain anything we like, while the URLs should be URL-encoded.

**Comment by nad** [comment]
> [...] while the URLs should be URL-encoded.

I have not checked things very carefully, but I think this amounts to percent-encoding the UTF8-encoded stream of octets. I'm not aware of any UTF8-encoders implemented in TeX/LaTeX...

**Comment by nad** [comment]
> I've tried porting the html backend to the blaze-html package

This package can generate HTML 5, right? Did you complete porting the backend?

**Comment by gallais** [comment]
I have nuked the branch because I thought it was useless. But it's a straightforward port so I can reproduce it if you think that'd be helpful.

**Comment by nad** [comment]
Yes, please do so.

**Comment by nad** [comment]
A summary of what I suggest that we do:
* [x] Switch to HTML 5.
* [x] Use the `id` attribute.
* [ ] Use **unique** Agda identifiers as the `id` attribute values, except that some characters (to be determined, at least including `&`) should be escaped in some way.
* [x] When we generate links to these identifiers we use the standard encoding (percent-encoding of the UTF-8 representation) of the values mentioned above as the target anchors.
* [ ] Users are free to use any kind of URLs in their documents. If they prefer to use some encoding scheme that is possibly not standards-compliant, then they can do this. We provide an Emacs procedure that performs the standard encoding (and perhaps also the escaping):
  ```emacs-lisp
  (defun agda2-encode-region-as-url (start end)
    "Encodes the region as a (hopefully) properly formatted URL."
    (interactive "r")
    (save-excursion
      (let ((s (url-encode-url (delete-and-extract-region start end))))
      (goto-char start)
      (insert s))))
  ```

**Comment by nad** [comment]
> some characters (to be determined, at least including `&`) should be escaped in some way.

* Ampersands [must not](https://www.w3.org/TR/2014/REC-html5-20141028/syntax.html#attributes-0) be [ambiguous](https://www.w3.org/TR/2014/REC-html5-20141028/syntax.html#syntax-ambiguous-ampersand). I suggest that we encode all ampersands using `&amp;`.
* If the value is delimited by double quotation marks (`"`), then double quotation marks inside the value must be encoded. However, we can't use double quotation marks in names.
* It does not appear as if `#`, `<`, `>` or `/` need to be escaped.

**Comment by nad** [comment]
> * When we generate links to these identifiers we use the standard encoding (percent-encoding of the UTF-8 representation) of the values mentioned above as the target anchors.
> * We provide an Emacs procedure that performs the standard encoding (and perhaps also the escaping)

I don't see a good reason for escaping the strings that should be encoded using what I called the standard encoding.

It would be nice to know that that the "standard encoding" is the right way to produce links that will resolve correctly by standards-compliant browsers. I have not studied the relevant standards in enough detail to be able to say for sure.

In Agda we distinguish between, say, `Ö` and `Ö` (where the former is a single code point and the latter consists of two code points). The HTML 5 specification [states](https://www.w3.org/TR/2014/REC-html5-20141028/infrastructure.html#syntax-references) that when a "hash-name reference" is compared to an `id` attribute value, then one should compare the strings code-point by code-point, like we do. (This does not apply to `name` attributes.)

I'd like to add that I don't know if file names with unusual characters are treated correctly. Perhaps we should do something about those as well. However, as far as I know @andreasabel did not change the way in which file names are treated, so this is perhaps less urgent.

**Comment by nad** [comment]
Perhaps non-ASCII code points will be allowed in the upcoming HTML 5 standard. The current draft [refers](http://w3c.github.io/html/infrastructure.html#urls-terminology) to the WHATWG URL standard, which [includes](https://url.spec.whatwg.org/#url-code-points) lots of code points. However, some kind of encoding may still be required, see below.

> Is there evidence for breakage?

Yes. Consider the following code:
```agda
module Bug where

A : Set₁
A = Set

%41 : Set₁
%41 = Set
```
Firefox 54.0 interprets the `%41` links as links to `A`. I think this is the correct interpretation. If you put the definition of `A` after the definition of `%41`, then Firefox interprets the links to `A` as links to `%41`. All of these links are displayed as `Bug.html#A` by Firefox.

**Comment by nad** [comment]
Actually, I think that it is already fine to use the WHATWG URL standard. There is also a [WHATWG HTML standard](https://html.spec.whatwg.org/multipage/) that some people seem to prefer to the W3C one, and that is perhaps closer to (certain?) major browser implementations.

**Comment by andreasabel** [comment]
In case you are waiting for me to fix this issue for the release #2659, I am not available soon enough.  Actually, @nad is now the most competent to fix this.

**Comment by nad** [comment]
Then I suggest that we roll back your changes (from the stable branch), rather than landing something that hasn't been tested properly just before the release.

**Comment by andreasabel** [comment]
Well, but even in its current state it is much more useful than what we had before.  I do not see a strong reason to roll it back.

**Comment by asr** [comment]
What about of releasing 2.5.3 based on the current stable-2.5 and if this issue is fixed *soon* then release 2.5.3.1 and deprecate 2.5.3 in Hackage?

**Comment by nad** [comment]
> Well, but even in its current state it is much more useful than what we had before.

As I demonstrated above the current implementation is broken. If you want to use the new implementation for some paper that you are preparing, then you can do this even if we don't release it just yet.

**Comment by nad** [comment]
I just opened the following issue: https://github.com/jaspervdj/blaze-html/issues/110.

**Comment by nad** [comment]
I found another problem: The same `name` is sometimes used for multiple nodes. Here is a cut-down example:
```agda
infix !_

postulate
  !_ : Set
```
Generated HTML:
```html
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"
><head
  ><title
    >Bug</title
    ><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"
     /><meta http-equiv="Content-Style-Type" content="text/css"
     /><link href="Agda.css" rel="stylesheet" type="text/css"
     /></head
  ><body
  ><pre
    ><a name="1" class="Keyword"
      >infix</a
      > <a name="7" class="Number"
      >0</a
      > <a name="!_" href="Bug.html#!_" class="Postulate Operator"
      >!_</a
      >

<a name="13" class="Keyword"
      >postulate</a
      >
  <a name="!_" href="Bug.html#!_" class="Postulate Operator"
      >!_</a
      > <a name="28" class="Symbol"
      >:</a
      > <a name="30" class="PrimitiveType"
      >Set</a
      >
</pre
    ></body
  ></html
>
```
Note that the two occurrences of `!_` have the same `name`.

It appears as if `defSiteHere` is [calculated](https://github.com/agda/agda/blob/867c083211e88c6890cd076d2b352da582e70539/src/full/Agda/Interaction/Highlighting/Generate.hs#L730) incorrectly for fixity declarations. @andreasabel, you added support for highlighting of operators in fixity declarations (#2140), do you know what's going on?

**Comment by nad** [comment]
Here is a related problem, that should perhaps be fixed at the same time. Consider the following code:
```agda
module %41 where

A : Set₁
A = Set
```
If we generate HTML for this file, then the generated links point to `A.html`, not `%41.html`.

**Comment by nad** [comment]
I've pushed an incomplete fix to the Issue2604 branch. The problem with fixity declarations has not been resolved.

**Comment by UlfNorell** [comment]
I fixed the fixity declaration problem (above commit). Does that mean we can merge the Issue2604 branch and close this issue?

**Comment by nad** [comment]
I guess so. Perhaps the issue numbers (#2604, #2725) should be added to the list of fixed issues in the changelog.

**Comment by asr** [comment]
I'll add the new issues fixed to the list of fixed issues in the CHANGELOG after freezing the branch.

**Comment by UlfNorell** [comment]
Ok. I'll go ahead and merge the branch.
